src/coreclr/vm/tieredcompilation.cpp

```diff
@@ -355,88 +392,190 @@ void WINAPI TieredCompilationManager::TieringDelayTimerCallback(PVOID parameter,
-                m_tieringDelayTimerHandle = nullptr;
+    int processorCount;
+#ifndef TARGET_UNIX
+    CPUGroupInfo::EnsureInitialized();
+    if (CPUGroupInfo::CanEnableGCCPUGroups() && CPUGroupInfo::CanEnableThreadUseAllCpuGroups())
+    {
+        processorCount = CPUGroupInfo::GetNumActiveProcessors();
+    }
+    else
+#endif
+    {
+        processorCount = GetCurrentProcessCpuCount();
+    }
+    _ASSERTE(processorCount > 0);
 
-                countOfNewMethodsCalledDuringDelay = m_countOfNewMethodsCalledDuringDelay;
-                m_countOfNewMethodsCalledDuringDelay = 0;
+    LARGE_INTEGER li;
+    QueryPerformanceFrequency(&li);
+    UINT64 ticksPerS = li.QuadPart;
+    UINT64 maxWorkDurationTicks = ticksPerS * 50 / 1000; // 50 ms
+    UINT64 minWorkDurationTicks = min(ticksPerS * processorCount / 1000, maxWorkDurationTicks); // <proc count> ms (capped)
+    UINT64 workDurationTicks = minWorkDurationTicks;
 
-                _ASSERTE(!IsTieringDelayActive());
+    while (true)
+    {
+        _ASSERTE(s_isBackgroundWorkerRunning);
+        _ASSERTE(s_isBackgroundWorkerProcessingWork);
 
-                if (!m_isBackgroundWorkScheduled && (m_isPendingCallCountingCompletion || m_countOfMethodsToOptimize != 0))
-                {
-                    m_isBackgroundWorkScheduled = true;
-                    doBackgroundWork = true;
-                }
+        if (IsTieringDelayActive())
+        {
+            do
+            {
+                ClrSleepEx(delayMs, false);
+            } while (!TryDeactivateTieringDelay());
+        }
 
-                break;
-            }
+        // Don't want to perform background work as soon as it is scheduled if there is possibly more important work that could
+        // be done. Some operating systems may also give a thread woken by a signal higher priority temporarily, which on a
+        // CPU-limited environment may lead to rejitting a method as soon as it's promoted, effectively in the foreground.
+        ClrSleepEx(0, false);
+
+        if (IsTieringDelayActive())
+        {
+            continue;
+        }
+
+        if ((m_isPendingCallCountingCompletion || m_countOfMethodsToOptimize != 0) &&
+            !DoBackgroundWork(&workDurationTicks, minWorkDurationTicks, maxWorkDurationTicks))
+        {
+            // Background work was interrupted due to the tiering delay being activated
+            _ASSERTE(IsTieringDelayActive());
+            continue;
         }
 
-        // Reschedule the timer if there has been recent tier 0 activity (when a new eligible method is called the first
-        // time) to further delay call counting
-        bool success = false;
-        EX_TRY
         {
-            if (ThreadpoolMgr::ChangeTimerQueueTimer(
-                    tieringDelayTimerHandle,
-                    g_pConfig->TieredCompilation_CallCountingDelayMs(),
-                    (DWORD)-1 /* Period, non-repeating */))
+            LockHolder tieredCompilationLockHolder;
```

Comment:
I think there is a little race window between exiting the lock inside of DoBackgroundWork() and re-taking the lock here but maybe I've misunderstood:

1. [Background thread] Inside DoBackgroundWork() within lock - determine all methods have been jitted
2. [Background thread] Return true from DoBackgroundWork() and then switch out before reaching this lock
3. [Managed code thread] Run AsyncPromoteToTier1 which calls TryScheduleBackgroundWorkerWithoutGCTrigger_Locked(). This call will observe that s_isBackgroundWorkerProcessingWork = true so s_backgroundWorkAvailableEvent is not set
4. [Background thread] Enter lock here, set s_isBackgroundWorkerProcessingWork = false, wait on event which may not be signaled again.

Reply:
The conditions for doing background work are rechecked inside the lock below, so if something happens in-between it would observe that inside the lock below and continue processing work.

> 4. [Background thread] Enter lock here, set s_isBackgroundWorkerProcessingWork = false

This part shouldn't happen because at step 3 `m_countOfMethodsToOptimize` would become nonzero and that would be seen below.

Reply:
Ah I think I follow now, this is a double checked lock pattern. the conditions evaluated above were just oportunistic and the precise checks are in the lock. Yep my mistake. I wouldn't have expected the performance needs of this code path to make that level of complexity worthwhile but if it matters no qualms from me.

Reply:
It wasn't particularly optimized, the only reason for the unlocked checks above was to avoid the unnecessary `BackgroundJitStart/Stop` events that would be fired by `DoBackgroundWork` when there is no background work to do.